### PR TITLE
Define Java Proxy and RayServeHandle

### PR DESCRIPTION
**Pull request by liuyang-my**
## Related issue number

This PR defines Java Proxy and RayServeHandle.

Detailed design is in doc :https://docs.google.com/document/d/1b9VCZsC2jVhWw_c_ifMu_ED3NxkARcOnU-OFqUYu-r4/edit#heading=h.b3vvz7iwnixn

The Python Proxy also needs some adjustments to call Java Backend. And We will commit a Java Proxy in next PR.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: java/serve/src/main/java/io/ray/serve/util/MetricsUtil.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  public static void registerMetrics(Runnable runnable) {
+    if (!Ray.isInitialized() || Ray.getRuntimeContext().isSingleProcess()) {
+      return;
+    }
+    runnable.run();
+  }
+
+  public static void xixihaha(Runnable runnable) {
+    runnable.run();
+  }
```
Comment (simon-mo) [inline java/serve/src/main/java/io/ray/serve/util/MetricsUtil.java:16]: remove

File: java/serve/src/test/java/io/ray/serve/HttpProxy.java [hunk 2]
```diff
@@ -0,0 +1,149 @@
+package io.ray.serve;
+
+import com.sun.net.httpserver.HttpExchange;
+import com.sun.net.httpserver.HttpHandler;
+import com.sun.net.httpserver.HttpServer;
```
Comment (simon-mo) [inline java/serve/src/test/java/io/ray/serve/HttpProxy.java:5]: what's the performance of `com.sun.net.httpserver.HttpServer`? It looks like it's has pretty bad performance, should we use a more performant Java HTTP server?

File: java/serve/src/main/java/io/ray/serve/RayServeHandle.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
+package io.ray.serve;
```
Comment (simon-mo) [inline java/serve/src/main/java/io/ray/serve/RayServeHandle.java:1]: need tests

File: java/serve/src/test/java/io/ray/serve/HttpProxy.java (hunk 2 above)
Comment (liuyang-my) [reply java/serve/src/test/java/io/ray/serve/HttpProxy.java:5]: Yes. The JDK HttpServer has two phase for each request, and it causes additional 40ms overhead in our product. Now we use another HTTP server instead JDK HttpServer. The default HTTP server of server proxy should be a high-performance one.

**Review by jiaodong — COMMENTED** [review]
just FYI: we're deprecating the term "backend" in serve and merging them back to "deployment" for consistency, including existing OSS java files (#18836), so while you're introducing new files it will be great to keep this in mind and don't diverge further :)

**Comment by liuyang-my** [comment]
> just FYI: we're deprecating the term "backend" in serve and merging them back to "deployment" for consistency, including existing OSS java files (#18836), so while you're introducing new files it will be great to keep this in mind and don't diverge further :)

OK, I'll pay attention about it.

File: java/serve/src/main/java/io/ray/serve/ReplicaSet.java [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    ActorHandle<RayServeWrappedReplica> replica =
+        handles.get(randomIndex); // TODO controll concurrency using maxConcurrentQueries
+    LOGGER.debug("Assigned query {} to replica {}.", query.getMetadata().getRequestId(), replica);
+    return replica
+        .task(RayServeWrappedReplica::handleRequest, query.getMetadata(), query.getArgs())
+        .remote();
+  }
+
+  public synchronized int
+      drainCompletedObjectRefs() { // TODO when using maxConcurrentQueries, change it to private.
```
Comment (simon-mo) [inline java/serve/src/main/java/io/ray/serve/ReplicaSet.java:138]: remove this or comment out because it's never called

**Review by simon-mo — COMMENTED** [review]
LGTM overall! Just one minor comment. Also the test failed: https://buildkite.com/ray-project/ray-builders-pr/builds/14821#5d0c1142-58d3-40b7-af78-2cb79c76cbf2/236-1214

**Review by simon-mo — APPROVED** [review]

File: java/serve/src/main/java/io/ray/serve/ReplicaSet.java (hunk 4 above)
Comment (liuyang-my) [reply java/serve/src/main/java/io/ray/serve/ReplicaSet.java:138]: OK.

**Review by simon-mo — APPROVED** [review]

**Review by jovany-wang — APPROVED** [review]

**Pull request merged**